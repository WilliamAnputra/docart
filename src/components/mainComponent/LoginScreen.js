import React, { Component } from 'react';
import {
  View,
  Image,
  TextInput,
  Text,
  BackHandler,
  Dimensions,
  KeyboardAvoidingView,
  StatusBar,
  PixelRatio,
  AsyncStorage,
  Keyboard,
  ToastAndroid,
  NetInfo
} from 'react-native';

import { Button } from 'react-native-elements';
import { NavigationActions } from 'react-navigation';
import Helper from '../../helper';

import * as constants from '../../constants/constants';
import * as globalColors from '../../constants/styles/global';
import * as icons from '../../constants/icons';


// List of Images
const SCREEN_WIDTH = Dimensions.get('window').width;
const SCREEN_HEIGHT = Dimensions.get('window').height;

const pixelRatio = PixelRatio.get();

export class Login extends Component {
  constructor(props) {
    super(props);

    this.state = {
      secureTextEntry: true,
      userName: '',
      password: '',
      textStart: 0,
      textEnd: 0,
      montserratRegular: '',
      montserratBold: '',
      emailOnFocus: false,
      passwordOnFocus: false
    };
  }

  componentDidMount() {
    BackHandler.addEventListener('hardwareBackPress', () => {
      BackHandler.exitApp();
      return true;
    });
  }
  componentWillUnmount() {
    BackHandler.removeEventListener('hardwareBackPress');
    return true;
  }

  // this is a function to dispatch login action
  async attemptLogin() {
    Keyboard.dismiss();

    // check if the network is stable
    NetInfo.getConnectionInfo().then((connectionInfo) => {

      // check if connection is un stable or offline
      if (connectionInfo.effectiveType == '2g' || connectionInfo.effectiveType == '3g' || connectionInfo.type == ' none') {
        return ToastAndroid.show('please check your connection', ToastAndroid.SHORT);
      }
    });

    // if there is a stable connection
    if (this.state.userName === '' || this.state.password === '') {
      return alert('Please fill in your email & password');
    }

    await AsyncStorage.setItem(constants.ACCESS_TOKEN, 'user has loggedIn');
    // this block of code clears any previous route
    // so our app doesn't go back to Login screen when user pressed back button
    const resetAction = NavigationActions.reset({
      index: 0,
      actions: [NavigationActions.navigate({ routeName: 'MainScreen' })]
    });

    return this.props.navigation.dispatch(resetAction);
  }

  render() {
    const {
      container,
      emailLabel,
      emailInput,
      separator,
      logo,
      loginButton
    } = styles;

    return (
      <View>
        <View style={container}>

          <KeyboardAvoidingView behavior="position" keyboardVerticalOffset={50}>

            <Image source={icons.LOGO} style={logo} />

            <Text
              style={[emailLabel, { fontFamily: this.state.montserratRegular }]}
            >
              EMAIL
            </Text>

            <TextInput
              placeholder="Youraccount@email.com"
              placeholderTextColor={globalColors.WHITE}
              style={[emailInput, { fontFamily: this.state.montserratRegular }]}
              underlineColorAndroid="transparent"
              onChangeText={userName => this.setState({ userName })}
              value={this.state.userName}
              onFocus={() => {
                this.setState({
                  emailOnFocus: true,
                  passwordOnFocus: false
                });
              }}
            />

            <View
              style={[
                separator,
                {
                  backgroundColor: this.state.emailOnFocus
                    ? globalColors.MEDIUM_GRAY
                    : globalColors.LIGHT_GRAY
                }
              ]}
            />

            <Text
              style={[emailLabel, { fontFamily: this.state.montserratRegular }]}
            >
              PASSWORD
            </Text>

            <View style={{ flexDirection: 'row' }}>
              <TextInput
                placeholder="********"
                placeholderTextColor={globalColors.WHITE}
                style={[
                  emailInput,
                  { fontFamily: this.state.montserratRegular }
                ]}
                onFocus={() => {
                  this.setState({
                    emailOnFocus: false,
                    passwordOnFocus: true
                  });
                }}
                secureTextEntry={this.state.secureTextEntry}
                underlineColorAndroid="transparent"
                onChangeText={password => this.setState({ password })}
                value={this.state.password}
                onSelectionChange={event => {
                  this.setState({
                    textEnd: event.nativeEvent.selection.end,
                    textStart: event.nativeEvent.selection.start
                  });
                }}
                selection={{
                  start: this.state.textStart,
                  end: this.state.textEnd
                }}
              />

            </View>

            <View
              style={[
                separator,
                {
                  backgroundColor: this.state.passwordOnFocus
                    ? globalColors.MEDIUM_GRAY
                    : globalColors.LIGHT_GRAY
                }
              ]}
            />

            <Button
              title="LOG IN"
              buttonStyle={loginButton}
              onPress={() => this.attemptLogin()}
            />

          </KeyboardAvoidingView>
        </View>
      </View>
    );
  }
}

const styles = {
  container: {
    height: SCREEN_HEIGHT,
    width: SCREEN_WIDTH,
    backgroundColor: globalColors.DARK_BLUE
  },
  emailLabel: {
    fontSize: 12,
    marginTop: Helper.scaleUnit(pixelRatio, 'height', 35),
    marginLeft: Helper.scaleUnit(pixelRatio, 'width', 35),
    backgroundColor: 'transparent',
    color: globalColors.WHITE
  },
  emailInput: {
    height: Helper.scaleUnit(pixelRatio, 'height', 35),
    width: Helper.scaleUnit(pixelRatio, 'width', 250),
    marginLeft: Helper.scaleUnit(pixelRatio, 'width', 35),
    opacity: 0.7,
    color: globalColors.WHITE
  },
  logo: {
    height: Helper.scaleUnit(pixelRatio, 'height', 80),
    width: Helper.scaleUnit(pixelRatio, 'width', 250),
    marginTop: Helper.scaleUnit(pixelRatio, 'height', 80),
    alignSelf: 'center',
  },
  loginButton: {
    width: Helper.scaleUnit(pixelRatio, 'width', 280),
    backgroundColor: globalColors.LIGHT_BLUE,
    height: SCREEN_HEIGHT * (50 / SCREEN_HEIGHT),
    alignItems: 'center',
    borderRadius: 4,
    marginTop: SCREEN_HEIGHT * (35 / SCREEN_HEIGHT),
    alignSelf: 'center'
  },
  separator: {
    width: Helper.scaleUnit(pixelRatio, 'width', 280),
    height: 2,
    marginLeft: Helper.scaleUnit(pixelRatio, 'width', 35),
    marginRight: Helper.scaleUnit(pixelRatio, 'width', 35)
  }
};

