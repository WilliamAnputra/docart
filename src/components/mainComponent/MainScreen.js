import React, { Component } from 'react';
import {
    View, Dimensions, PixelRatio, TouchableOpacity, TouchableWithoutFeedback,
    Text, FlatList, Image, ToastAndroid, BackHandler, AsyncStorage
} from 'react-native';

import { RNCamera } from 'react-native-camera';
import ImagePicker from 'react-native-image-picker';

import Icon from 'react-native-vector-icons/Ionicons';

import Helper from '../../helper';
import * as constants from '../../constants/constants';
import * as globalColors from '../../constants/styles/global';
import * as icons from '../../constants/icons';


import { AttachmentAlertBox } from '../subComponent/AttachmentAlertBox';


// List of Images
const SCREEN_WIDTH = Dimensions.get('window').width;
const SCREEN_HEIGHT = Dimensions.get('window').height;

const photoLimitWarning = 'Anda sudah mencapai batas jumlah upload foto';

const pixelRatio = PixelRatio.get();

export class MainScreen extends Component {

    constructor(props) {
        super(props);
        this.state = {
            photos: [],
            hideOption: true,
            showCamera: false,
            showGallery: false,
        };
    }

    async componentWillMount() {
        // check if we have any store photos
        const savedPhotos = await AsyncStorage.getItem(constants.PHOTOS);

        if (savedPhotos == null) {
            return null;
        }

        // if we do have photos then use our current photo
        return this.setState({ photos: JSON.parse(savedPhotos) });
    }

    componentDidMount() {

        BackHandler.addEventListener('hardwareBackPress', () => {
            return BackHandler.exitApp();
        });
    }

    componentWillUnmount() {
        return BackHandler.removeEventListener('hardwareBackPress');
    }


    // function when user press 'Snap' button
    takePicture = async () => {

        if (this.camera) {
            const options = { quality: 1, base64: true, width: 600, exif: true };
            const data = await this.camera.takePictureAsync(options);
            await this.setState({ photos: [...this.state.photos, data.uri], showCamera: false, hideOption: true }, () => {
                AsyncStorage.setItem(constants.PHOTOS, JSON.stringify(this.state.photos));
            });

        }
    };

    // a container to show or hide 'camera' preview screen
    showCamera() {
        if (this.state.showCamera) {
            return (
                <View style={{
                    flexDirection: 'column',
                    backgroundColor: 'black',
                    position: 'absolute',
                    height: SCREEN_HEIGHT,
                    width: SCREEN_WIDTH
                }}>
                    <RNCamera
                        ref={ref => {
                            this.camera = ref;
                        }}
                        style={styles.preview}
                        type={RNCamera.Constants.Type.back}
                        flashMode={RNCamera.Constants.FlashMode.off}
                        permissionDialogTitle={'Permission to use camera'}
                        permissionDialogMessage={'We need your permission to use your camera phone'}
                    />
                    <View style={{ flex: 0, flexDirection: 'row', justifyContent: 'center', }}>
                        <TouchableOpacity
                            onPress={() => this.takePicture()}
                            style={styles.capture}
                        >
                            <Text style={{ fontSize: 14 }}> SNAP </Text>
                        </TouchableOpacity>

                        <TouchableOpacity
                            onPress={async () => this.setState({ showCamera: false, hideOption: true })}
                            style={styles.capture}
                        >
                            <Text style={{ fontSize: 14 }}> Close </Text>
                        </TouchableOpacity>
                    </View>
                </View>
            );
        }
        return null;

    }

    // floating button component
    async floatingButtonOnPress() {
        if (this.state.photos.length == 3) {
            return ToastAndroid.show(photoLimitWarning, ToastAndroid.SHORT);
        }

        return this.setState({ hideOption: false });
    }

    showGallery() {
        if (this.state.showGallery) {
            const options = {
                maxWidth: 600,
                maxHeight: 450,
            };
            // Open Image Library:
            ImagePicker.launchImageLibrary(options, (response) => {

                // if user cancels
                if (response.didCancel == true) {
                    return this.setState({ showGallery: false, hideOption: true });
                }

                const imageUri = response.uri;
                this.setState({ showGallery: false, hideOption: true, photos: [...this.state.photos, imageUri] }, () => {
                    AsyncStorage.setItem(constants.PHOTOS, JSON.stringify(this.state.photos));
                });
            });
        }
        return null;
    }

    // callback function to be passed to 'AttachmentAlertBox' as a callback function
    onCameraPress = async () => {
        await this.setState({ showCamera: true });
    }
    // callback function to be passed to 'AttachmentAlertBox' as a callback function
    onGalleryPress = async () => {
        await this.setState({ showGallery: true });
    }

    // function to show photo upload method to user
    showUploadOptions() {
        if (this.state.hideOption == false) {
            return <AttachmentAlertBox onCameraPress={this.onCameraPress} onGalleryPress={this.onGalleryPress} />;
        }
        return null;
    }

    // flatlist's automatic detect prop if data is empty
    renderEmptyList() {
        return (
            <Text style={{ alignSelf: 'center' }}> No photos available</Text>
        );
    }

    // function to delete photos
    async deletePhoto() {
        await this.setState({ photos: [] });
    }

    async logOut() {
        AsyncStorage.multiRemove([constants.ACCESS_TOKEN, constants.PHOTOS], (err) => {
            if (err) {
                return ToastAndroid.show(err, ToastAndroid.SHORT);
            }
        });

        return this.props.navigation.navigate('LoginScreen');
    }

    render() {
        const DismissBox = ({ children }) => {
            return (
                <TouchableWithoutFeedback onPress={() => this.setState({ hideOption: true })}>
                    {children}
                </TouchableWithoutFeedback>
            );
        };

        const { container, toolbar, logo, logOutButton, imageSize, deletePhotosBox, floatingActionButton } = styles;

        return (
            <DismissBox>
                <View style={container}>
                    <View style={toolbar} />
                    <Image source={icons.LOGO} style={logo} />

                    <TouchableOpacity
                        style={logOutButton}
                        onPress={() => this.logOut()}>

                        <Text style={{ color: globalColors.WHITE, marginTop: 5 }}> LogOut </Text>
                    </TouchableOpacity>

                    <FlatList
                        data={this.state.photos}
                        style={{ height: SCREEN_HEIGHT }}
                        numColumns={2}
                        keyExtractor={data => data}
                        extraData={this.state.photos}
                        ListEmptyComponent={() => this.renderEmptyList()}
                        renderItem={data => {
                            return (
                                <Image source={{ uri: data.item }} style={imageSize} />
                            );
                        }}
                    />

                    <TouchableOpacity
                        style={deletePhotosBox}
                        onPress={() => this.deletePhoto()}>

                        <Text style={{ color: globalColors.WHITE }}> Delete </Text>
                    </TouchableOpacity>

                    <TouchableOpacity
                        style={floatingActionButton}
                        onPress={() => this.floatingButtonOnPress()}>

                        <Icon name="md-add" size={30} color={globalColors.WHITE} />

                    </TouchableOpacity>



                    {this.showUploadOptions()}

                    {this.showCamera()}

                    {this.showGallery()}
                </View>

            </DismissBox>

        );
    }
}


const styles = {
    container: {
        flex: 1,
        backgroundColor: globalColors.WHITE
    },
    backArrow: {
        position: 'absolute',
        top: Helper.scaleUnit(pixelRatio, 'height', 10),
        left: Helper.scaleUnit(pixelRatio, 'height', 10),
    },
    capture: {
        flex: 0,
        backgroundColor: globalColors.WHITE,
        borderRadius: 5,
        padding: Helper.scaleUnit(pixelRatio, 'height', 15),
        paddingHorizontal: Helper.scaleUnit(pixelRatio, 'width', 20),
        alignSelf: 'center',
        margin: Helper.scaleUnit(pixelRatio, 'height', 20),
    },
    floatingActionButton: {
        position: 'absolute',
        right: Helper.scaleUnit(pixelRatio, 'width', 10),
        bottom: Helper.scaleUnit(pixelRatio, 'height', 40),
        width: Helper.scaleUnit(pixelRatio, 'width', 50),
        height: Helper.scaleUnit(pixelRatio, 'height', 50),
        borderRadius: 25,
        borderColor: 'transparent',
        borderWidth: 0.4,
        backgroundColor: globalColors.DARK_BLUE,
        justifyContent: 'center',
        alignItems: 'center'
    },
    deletePhotosBox: {
        position: 'absolute',
        left: Helper.scaleUnit(pixelRatio, 'width', 10),
        bottom: Helper.scaleUnit(pixelRatio, 'height', 40),
        width: Helper.scaleUnit(pixelRatio, 'width', 100),
        height: Helper.scaleUnit(pixelRatio, 'height', 50),
        borderColor: 'transparent',
        borderWidth: 0.4,
        backgroundColor: globalColors.DARK_BLUE,
        justifyContent: 'center',
        alignItems: 'center'
    },
    imageSize: {
        height: Helper.scaleUnit(pixelRatio, 'height', 200),
        width: Helper.scaleUnit(pixelRatio, 'height', 200),
        margin: 5
    },
    logo: {
        height: Helper.scaleUnit(pixelRatio, 'height', 30),
        width: Helper.scaleUnit(pixelRatio, 'width', 100),
        alignSelf: 'center',
        position: 'absolute',
        top: Helper.scaleUnit(pixelRatio, 'height', 10),
    },
    logOutButton: {
        position: 'absolute',
        top: 0,
        right: 10
    },
    preview: {
        flex: 1,
        justifyContent: 'flex-end',
        alignItems: 'center'
    },
    toolbar: {
        height: Helper.scaleUnit(pixelRatio, 'height', 56),
        width: SCREEN_WIDTH,
        backgroundColor: globalColors.DARK_BLUE
    },
};
