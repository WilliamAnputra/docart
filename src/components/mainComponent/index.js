export * from './LoginScreen';
export * from './MainScreen';
