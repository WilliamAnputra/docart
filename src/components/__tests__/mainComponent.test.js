import React from 'react';
import renderer from 'react-test-renderer';
import { View } from 'react-native';

import { Login } from '../mainComponent/LoginScreen';

it('Renders login screen component', () => {
    const rendered = renderer.create(<Login> <View /> </Login>).toJSON();
    expect(rendered).toMatchSnapshot();
});


