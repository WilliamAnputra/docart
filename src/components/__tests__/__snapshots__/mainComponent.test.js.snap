// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders login screen component 1`] = `
<View>
  <View
    style={
      Object {
        "backgroundColor": "#253f61",
        "height": 1334,
        "width": 750,
      }
    }
  >
    <View
      keyboardVerticalOffset={50}
      onLayout={[Function]}
      style={undefined}
    >
      <View
        style={
          Array [
            undefined,
            Object {
              "bottom": 0,
            },
          ]
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/images/logo.png",
            }
          }
          style={
            Object {
              "alignSelf": "center",
              "height": 80,
              "marginTop": 80,
              "width": 250,
            }
          }
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "color": "#ffffff",
                "fontSize": 12,
                "marginLeft": 35,
                "marginTop": 35,
              },
              Object {
                "fontFamily": "",
              },
            ]
          }
        >
          EMAIL
        </Text>
        <TextInput
          allowFontScaling={true}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Youraccount@email.com"
          placeholderTextColor="#ffffff"
          style={
            Array [
              Object {
                "color": "#ffffff",
                "height": 35,
                "marginLeft": 35,
                "opacity": 0.7,
                "width": 250,
              },
              Object {
                "fontFamily": "",
              },
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
        <View
          style={
            Array [
              Object {
                "height": 2,
                "marginLeft": 35,
                "marginRight": 35,
                "width": 280,
              },
              Object {
                "backgroundColor": "#cccccc",
              },
            ]
          }
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
                "color": "#ffffff",
                "fontSize": 12,
                "marginLeft": 35,
                "marginTop": 35,
              },
              Object {
                "fontFamily": "",
              },
            ]
          }
        >
          PASSWORD
        </Text>
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            onChangeText={[Function]}
            onFocus={[Function]}
            onSelectionChange={[Function]}
            placeholder="********"
            placeholderTextColor="#ffffff"
            secureTextEntry={true}
            selection={
              Object {
                "end": 0,
                "start": 0,
              }
            }
            style={
              Array [
                Object {
                  "color": "#ffffff",
                  "height": 35,
                  "marginLeft": 35,
                  "opacity": 0.7,
                  "width": 250,
                },
                Object {
                  "fontFamily": "",
                },
              ]
            }
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View
          style={
            Array [
              Object {
                "height": 2,
                "marginLeft": 35,
                "marginRight": 35,
                "width": 280,
              },
              Object {
                "backgroundColor": "#cccccc",
              },
            ]
          }
        />
        <View
          style={
            Array [
              Object {
                "marginLeft": 15,
                "marginRight": 15,
              },
              undefined,
              undefined,
              undefined,
            ]
          }
        >
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            hasTVPreferredFocus={undefined}
            hitSlop={undefined}
            isTVSelectable={true}
            nativeID={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
            testID={undefined}
            tvParallaxProperties={undefined}
          >
            <View
              pointerEvents="box-only"
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#9E9E9E",
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "padding": 19,
                  },
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  Object {
                    "padding": 12,
                  },
                  undefined,
                  undefined,
                  undefined,
                  Object {
                    "alignItems": "center",
                    "alignSelf": "center",
                    "backgroundColor": "#4A90E2",
                    "borderRadius": 4,
                    "height": 50,
                    "marginTop": 35,
                    "width": 280,
                  },
                  undefined,
                  undefined,
                ]
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {},
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                    Array [
                      Object {
                        "color": "white",
                        "fontSize": 20,
                      },
                      undefined,
                      Object {
                        "fontSize": 17.5,
                      },
                      undefined,
                      undefined,
                      undefined,
                      undefined,
                      undefined,
                    ],
                  ]
                }
              >
                LOG IN
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
